README.md
# 🛫 RDGliderViewController 🛬

[![Build Status](https://travis-ci.org/gelemias/RDGliderViewController.svg?branch=develop)](https://travis-ci.org/gelemias/RDGliderViewController) [![Coverage Status](https://coveralls.io/repos/github/gelemias/RDGliderViewController/badge.svg?branch=develop)](https://coveralls.io/github/gelemias/RDGliderViewController?branch=develop) [![License: MIT](https://img.shields.io/github/license/mashape/apistatus.svg)](https://opensource.org/licenses/MIT) [![Carthage compatible](https://img.shields.io/badge/Carthage-compatible-4BC51D.svg?style=flat)](https://github.com/Carthage/Carthage)


Control in *Objective-C* for a draggable ViewController gliding over another ViewController. 

**⚠️⚠️ This project is discontinued, use [RDGliderViewController-Swift](https://github.com/gelemias/RDGliderViewController-Swift)  instead ⚠️⚠️**

RDGliderViewController is a view controller that manages a scrollable view using one side as a sliding point, choosing between Left-to-Right, Top-to-Bottom, Right-to-Left and Bottom-to-Top, and setting an array of offsets as percent values to determine the gliding view steps, as described bellow.

<img src="./img/1.gif" width="180"> <img src="./img/2.gif" width="180"> <img src="./img/3.gif" width="180"> <img src="./img/4.gif" width="180">

## Example

A good way to learn how to use RDGliderViewController is to go through the example app. Take a look at the tests as well for a more thorough usage.

To run the example project, clone the repo, and open `./Example/GliderSample.xcodeproj`

## Installation

### Using CocoaPods
RDGliderViewController is available through [CocoaPods](http://cocoapods.org), a dependency manager for Swift and Objective-C Cocoa projects. To install
it, simply add the following line to your [Podfile](https://guides.cocoapods.org/using/getting-started.html):

```ruby
pod 'RDGliderViewController'
```

### Using Carthage

[Carthage](https://github.com/Carthage/Carthage) is a decentralized dependency manager that builds your dependencies and provides you with binary frameworks.

You can install Carthage with [Homebrew](http://brew.sh/) and then integrate **RDGliderViewController** into your Xcode project by adding the following line to your [Cartfile](https://github.com/Carthage/Carthage/blob/master/Documentation/Artifacts.md#cartfile):

```ruby
github "gelemias/RDGliderViewController"
```

### Manual Installation

For an old fashion installation you can directly add the **header** and **implementation** files for `RDGliderViewController` and `RDScrollView` to your project.

- Download the latest code version or add the repository as a git submodule to your git-tracked project.

- Drag and drop the folder `RDGliderViewController/` onto your project and make sure to select **Copy items if needed**.

- Add `#import "RDGliderViewController.h"` at the beginning of your class.

## Usage

RDGliderViewController is a very simple to use controller that mainly depends on three values to work on, *Orientation type* which is determines the scrolling side, the *List of Offsets* which represent the steps of expansion and contraction and the content itself with is a view controller completely independent.

### Creation

Use a the following custom init to instantiate RDGliderViewController:

`- (instancetype)initOn:(nonnull UIViewController *)parent WithContent:(nonnull RDGliderContentViewController *)content type:(RDScrollViewOrientationType)type AndOffsets:(nonnull NSArray<NSNumber *> *)offsets {
`

e.g.
```Objective-C
RDGliderViewController *gliderVC = [[RDGliderViewController alloc] initOn:self
                                                              WithContent:[UIViewController new]
                                                                     type:RDScrollViewOrientationRightToLeft
                                                               AndOffsets:@[@0.2, @0.5, @1]];
```

And that's all you need to do, if you run your app, an instance of UIViewController should be scrollable from Right to Left on top of `self` view controller.

### Content and Container

The content view controller should inherit from RDGliderContentViewController to inherit the properly resize when rotating, is treated as a child view controller of this one.

```Objective-C
#import "RDGliderContentViewController.h"

@interface ContentViewController : RDGliderContentViewController

// Your content class Here

@end

```

Mentioned that not much more is required, only keep in mind that if you do not define a size for the view, RDGliderViewController will it resize it to match container's.

Content view controller can be afterwards as well but always along offsets and orientation since these three properties are dependent of each other:

`- (void)setContentViewController:(nonnull RDGliderContentViewController *)contentViewController
                            type:(RDScrollViewOrientationType)type
                         offsets:(nonnull NSArray<NSNumber *> *)offsets;
`

### Orientation types

`RDScrollViewOrientationType` represent the four sides of the display plus the sliding direction.

```Objective-C
typedef NS_ENUM(unsigned int, RDScrollViewOrientationType) {
    RDScrollViewOrientationUnknown = 0,

    RDScrollViewOrientationLeftToRight,
    RDScrollViewOrientationBottomToTop,
    RDScrollViewOrientationRightToLeft,
    RDScrollViewOrientationTopToBottom

};
```

if content view controller defines a fixed size, this size will be respected during rotation, for horizontal scrolls (Left-to-Right and Right-to-Left) will resize *Width* and for Vertical (Bottom-to-Top and Top-to-Bottom) will reize *Height*


### Offsets and OffsetIndex

this `NSArray<NSNumber *> *offsets` property is a list of steps used when either expanding / collapsing or dragging the list is represented in percent % related with the length of the content view controller, e.g. given the list [0, 0.5, 1] for a lenght *400px* would be this offset steps of [0px, 200px, 400px]

`offsets[OffsetIndex].floatValue * contentViewController.lenght`

For inverted OrientationTypes (RDScrollViewOrientationLeftToRight and RDScrollViewOrientationTopToBottom), the offset list are also inverted so for example for a list [0.2, 0.4, 0.6, 1] will be returned [0.8, 0.6, 0.4, 0]

### Expand, Collapse and Close

By default RDGliderViewController is draggable but when setting an offset of `0.0f%` it will be outside of the screen on starting point, to Increase the position of the view by one in the list of offsets use:

`- (void)expand;`

to Decrease the position of RDGliderViewController by one in the list of offsets:

`- (void)collapse;`

and *close* which moves the View directly to the first offset which is the default position:

`- (void)close;`

All of these methods simple change the offsetIndex when calling this method.

`- (void)changeOffsetTo:(NSUInteger)offsetIndex animated:(BOOL)animated;`

## License

RDGliderViewController is released under the MIT license. See [LICENSE](https://github.com/gelemias/RDGliderViewController/blob/develop/LICENSE) for details.


## Change log

See [CHANGELOG](https://github.com/gelemias/RDGliderViewController/blob/develop/CHANGELOG.md) for details.


CHANGELOG.md
## CHANGELOG

#### Latest version -  0.1.76
+ 2017-11-22 *[FIX] Fixed issue happening with changing margin and starting from different offset than 0*
+ 2017-05-02 *[ENV] Added a before_install instruction to travis yml*
+ 2017-05-02 *[ENV] Updated ruby version required in gem file*
+ 2017-04-27 *[REF] refactored Scrollview to be aligned with the Swift project*

#### Previously released:
+ 2017-04-21 *[REF] created a base RDGliderContentViewController meant to be the base case for any content view controller*
+ 2017-04-21 *[FIX] Fixed hidden behaviour with inverted orientationTypes*
+ 2017-04-21 *[POD] updated podspec*
+ 2017-04-21 *[DOC] Completed a basic documentation to the README and added a CHANGELOG*
+ 2017-04-20 *[DOC] Update README with CocoaPods and Carthage support*
+ 2017-04-20 *[ENV] preparing project for a pod lib*
+ 2017-04-18 *[ENV] Added code coverage support*
+ 2017-04-18 *[ENV] added yml build file for Travis CI*
+ 2017-04-18 *[TEST] Added unit tests for GVScrollView*
+ 2017-04-14 *[REF] enhanced  naming*
+ 2017-04-14 *[FIX] issue with rendering shadow when rotating*
+ 2017-04-14 *[FIX] horizontal offset broken*
+ 2017-04-13 *[DEV] first implementation of Left To Right side*
+ 2017-04-13 *[REF] cleaned constraints of scrollview*
+ 2017-04-13 *[DEV] limited max content scroll to largest offset percent of content length*
+ 2017-04-13 *[DEV] added margins to reversed positions*
+ 2017-04-13 *[DEV] first implementation for TopToBottom GlideView*
+ 2017-04-12 *[FIX] fixed release notes not recognised characters*
+ 2017-04-12 *[DOC] added ReleaseNotes to Readme file*
+ 2017-04-07 *[DEV] enhanced threshold calculation for boundaries*
+ 2017-04-07 *[DOC] enhanced documentation on headers*
+ 2017-04-07 *[DEV] improved constructor of GlideViewController*
+ 2017-04-07 *[FIX] offset changes where altered while rotating*
+ 2017-04-06 *[REF] refactored contentSize handling when rotating*
+ 2017-04-06 *[FIX] offset is affected by the margin when rotating a rather large view*
+ 2017-04-06 *[DEV] enable the possibility to change multiple offsets by dragging*
+ 2017-04-06 *[REF] changed offsets approach to % dependant on content’s view*
+ 2017-04-06 *[REF] refactor usage of glideView types*
+ 2017-04-05 *[DEV] enhanced recalculation of offset when rotating scrollview*
+ 2017-04-05 *[FIX] landscape margins offset is off*
+ 2017-04-05 *[FIX] landscape to portrait rotation changes offset of scrollview*
+ 2017-04-05 *[FIX] landscape margins offset is off*


Example/GliderSample/ViewController.m
//
//  ViewController.m
//  GliderSample
//
//  Created by Guillermo Delgado on 03/04/2017.
//  Copyright © 2017. All rights reserved.
//

#import "ViewController.h"
#import "RDGliderViewController.h"
#import "ContentViewController.h"
#import "RDGradientView.h"

@interface ViewController () <RDGliderViewControllerDelegate>

@property (nonatomic) RDGliderViewController *leftToRightGlideVC;
@property (nonatomic) RDGliderViewController *bottomToTopGlideVC;
@property (nonatomic) RDGliderViewController *topToBottomGlideVC;
@property (nonatomic) RDGliderViewController *rightToLeftGlideVC;

@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    RDGradientView *bgGradient = [[RDGradientView alloc] initWithFrame:self.view.bounds];
    bgGradient.autoresizingMask = UIViewAutoresizingFlexibleWidth | UIViewAutoresizingFlexibleHeight;
    bgGradient.layer.colors = @[(id)[UIColor colorWithRed:(CGFloat)0.643 green:(CGFloat)0.569 blue:(CGFloat)0.776 alpha:1].CGColor,
                                     (id)[UIColor colorWithRed:(CGFloat)(CGFloat)0.573 green:(CGFloat)0.875 blue:(CGFloat)0.678 alpha:1].CGColor];
    [self.view insertSubview:bgGradient atIndex:0];
    
    [self initRightToLeftGlideView];
    [self initBottomToTopGlideView];
    [self initTopToBottomGlideView];
    [self initLeftToRightGlideView];
}

- (void)initRightToLeftGlideView{
    self.rightToLeftGlideVC = [[RDGliderViewController alloc] initOn:self
                                                      WithContent:[[ContentViewController alloc] initWithLength:200.0f]
                                                             type:RDScrollViewOrientationRightToLeft
                                                       AndOffsets:@[@(0), @(1)]];
    self.rightToLeftGlideVC.delegate = self;
}

- (void)initBottomToTopGlideView {
    self.bottomToTopGlideVC = [[RDGliderViewController alloc] initOn:self
                                                      WithContent:[ContentViewController new]
                                                             type:RDScrollViewOrientationBottomToTop
                                                       AndOffsets:@[@(0),
                                                                    @(0.6),
                                                                    @(0.2),
                                                                    @(0.4),
                                                                    @(0.8)]];
    self.bottomToTopGlideVC.delegate = self;
    self.bottomToTopGlideVC.marginOffset = 30;
}

- (void)initTopToBottomGlideView {
    self.topToBottomGlideVC = [[RDGliderViewController alloc] initOn:self
                                                      WithContent:[[ContentViewController alloc] initWithLength:400.0f]
                                                             type:RDScrollViewOrientationTopToBottom
                                                       AndOffsets:@[@(0),
                                                                    @(0.5),
                                                                    @(1)]];
    self.topToBottomGlideVC.delegate = self;
}

- (void)initLeftToRightGlideView {
    self.leftToRightGlideVC = [[RDGliderViewController alloc] initOn:self
                                                      WithContent:[ContentViewController new]
                                                             type:RDScrollViewOrientationLeftToRight
                                                       AndOffsets:@[@(0),
                                                                    @(0.6),
                                                                    @(0.2),
                                                                    @(0.4),
                                                                    @(0.8),
                                                                    @(1)]];
    self.leftToRightGlideVC.delegate = self;
    self.leftToRightGlideVC.marginOffset = 10;
}

#pragma mark - Actions

- (IBAction)bottomToTopButtonPressed:(id)sender {
    if ([self.bottomToTopGlideVC currentOffsetIndex] < [[self.bottomToTopGlideVC offsets] count] - 1) {
        [self.bottomToTopGlideVC expand];
    } else {
        [self.bottomToTopGlideVC shake];
    }
}

- (IBAction)leftToRightButtonPressed:(id)sender {
    if ([self.leftToRightGlideVC currentOffsetIndex] < [[self.leftToRightGlideVC offsets] count] - 1) {
        [self.leftToRightGlideVC expand];
    } else {
        [self.leftToRightGlideVC shake];
    }
}

- (IBAction)topToBottomButtonPressed:(id)sender {
    if ([self.topToBottomGlideVC currentOffsetIndex] < [[self.topToBottomGlideVC offsets] count] - 1) {
        [self.topToBottomGlideVC expand];
    } else {
        [self.topToBottomGlideVC shake];
    }
}

- (IBAction)rightToLeftButtonPressed:(id)sender {
    if ([self.rightToLeftGlideVC currentOffsetIndex] < [[self.rightToLeftGlideVC offsets] count] - 1) {
        [self.rightToLeftGlideVC expand];
    } else {
        [self.rightToLeftGlideVC shake];
    }
}

#pragma mark - GlideViewControllerDelegate

- (void)glideViewController:(RDGliderViewController *)glideViewController hasChangedOffsetOfContent:(CGPoint)offset {
    ContentViewController *vc = (ContentViewController *)glideViewController.contentViewController;
    [vc setOffset:NSStringFromCGPoint(offset)];
}

- (void)glideViewControllerDidExpand:(RDGliderViewController *)glideViewController {
    ContentViewController *vc = (ContentViewController *)glideViewController.contentViewController;
    [vc setIndex:glideViewController.currentOffsetIndex ofMax:[glideViewController.offsets count] - 1];
}

- (void)glideViewControllerDidCollapse:(RDGliderViewController *)glideViewController {
    ContentViewController *vc = (ContentViewController *)glideViewController.contentViewController;
    [vc setIndex:glideViewController.currentOffsetIndex ofMax:[glideViewController.offsets count] - 1];
}

@end
